File: perf-results/cross-lang/benchmarks/fannkuchredux.lua
local function fannkuch(n)
  local perm1 = {}
  local count = {}
  for i = 0, n-1 do perm1[i] = i end
  for i = 0, n-1 do count[i] = i + 1 end

  local max_flips = 0
  local checksum = 0
  local sign = 1

  while true do
    if perm1[0] ~= 0 then
      local perm = {}
      for i = 0, n-1 do perm[i] = perm1[i] end
      local flips = 0
      local k = perm[0]
      while k ~= 0 do
        local lo, hi = 0, k
        while lo < hi do
          perm[lo], perm[hi] = perm[hi], perm[lo]
          lo = lo + 1; hi = hi - 1
        end
        flips = flips + 1
        k = perm[0]
      end
      if flips > max_flips then max_flips = flips end
      checksum = checksum + sign * flips
    end

    sign = -sign
    local i = 1
    while i < n do
      local v = perm1[i]
      for j = i, 1, -1 do perm1[j] = perm1[j-1] end
      perm1[0] = v
      count[i] = count[i] - 1
      if count[i] > 0 then break end
      count[i] = i + 1
      i = i + 1
    end
    if i >= n then break end
  end

  return max_flips, checksum
end

local N = tonumber(arg and arg[1]) or 7
local max_flips, checksum = fannkuch(N)
io.write(string.format("%d\nPfannkuchen(%d) = %d\n", checksum, N, max_flips))

File: perf-results/cross-lang/benchmarks/fannkuchredux.rb
def fannkuch(n)
  perm1 = (0...n).to_a
  count = (1..n).to_a
  max_flips = 0
  checksum = 0
  sign = 1

  loop do
    if perm1[0] != 0
      perm = perm1.dup
      flips = 0
      k = perm[0]
      while k != 0
        perm[0..k] = perm[0..k].reverse
        flips += 1
        k = perm[0]
      end
      max_flips = flips if flips > max_flips
      checksum += sign * flips
    end

    sign = -sign
    i = 1
    while i < n
      perm1.insert(0, perm1.delete_at(i))
      count[i] -= 1
      break if count[i] > 0
      count[i] = i + 1
      i += 1
    end
    break if i >= n
  end

  [max_flips, checksum]
end

n = (ARGV[0] || 7).to_i
max_flips, checksum = fannkuch(n)
puts "#{checksum}\nPfannkuchen(#{n}) = #{max_flips}"
